### PR TITLE
Message, not throw, failures in WEA.startRunnableScopes.

### PR DESCRIPTION
**Pull request by kshakir**
Other syntax fixes.

### CODE REVIEW

**Comment by kshakir** [comment]
Centaur PR https://github.com/broadinstitute/centaur/pull/164

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala [hunk 1]
```diff
@@ -355,34 +359,36 @@ case class WorkflowExecutionActor(workflowDescriptor: EngineWorkflowDescriptor,
...
+        case k: DynamicDeclarationKey => processRunnableDynamicDeclaration(k, data)
+        case k => Failure(new UnsupportedOperationException(s"Unknown entry in execution store: ${k.tag}"))
+      }).flatten
+    } map {
+      case (_, Success(value)) => Success(value)
+      case (scope, Failure(throwable)) =>
+        self ! JobInitializationFailed(scope, throwable)
+        Failure(throwable)
+    } collect {
+      case Success(value) => value
```
Comment (Horneth) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: TOL: Doesn't that mean that we're potentially going to start new jobs even though the workflow is doomed to fail if one of the `JobInitializationFailed` message was sent ?

Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: Great catch!  It looks like we still want the `sequence` behavior here.

File: .travis.yml [hunk 2]
```diff
@@ -19,7 +19,7 @@ before_install:
   - openssl aes-256-cbc -K "$encrypted_5ebd3ff04788_key" -iv "$encrypted_5ebd3ff04788_iv" -in src/bin/travis/resources/jesConf.tar.enc -out jesConf.tar -d
 env:
   global:
-    - CENTAUR_BRANCH=develop
+    - CENTAUR_BRANCH=ks_catch_runnable_scopes
```
Comment (cjllanwarne) [inline .travis.yml]: Remember to revert before merging!

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala [hunk 3]
```diff
@@ -259,30 +259,34 @@ case class WorkflowExecutionActor(workflowDescriptor: EngineWorkflowDescriptor,
...
-      data.expressionLanguageFunctions,
-      data.outputStore.fetchNodeOutputEntries
-    ) map { workflowOutputs =>
-       // For logging and metadata
-       val workflowScopeOutputs = workflowOutputs map {
-         case (output, value) => output.locallyQualifiedName(workflowDescriptor.workflow) -> value
-       }
-       workflowLogger.info(
-         s"""Workflow ${workflowDescriptor.workflow.unqualifiedName} complete. Final Outputs:
+    val responseAndFinalStateTry: Try[(WorkflowExecutionActorResponse, WorkflowExecutionActorState)] =
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: Channelling the honourable @geoffjentry, can we ADT this?

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala [hunk 4]
```diff
@@ -259,30 +259,34 @@ case class WorkflowExecutionActor(workflowDescriptor: EngineWorkflowDescriptor,
...
+          case (output, value) => output.unqualifiedName -> JobOutput(value)
+        }
+        (WorkflowExecutionSucceededResponse(data.jobExecutionMap, unqualifiedWorkflowOutputs),
+          WorkflowExecutionSuccessfulState)
+      }
+
+    val (response, finalState) = responseAndFinalStateTry recover {
+      case throwable =>
+        (WorkflowExecutionFailedResponse(data.jobExecutionMap, throwable), WorkflowExecutionFailedState)
     } get
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: ToL: I was happy right up until I saw this!

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala [hunk 5]
```diff
@@ -493,7 +499,7 @@ case class WorkflowExecutionActor(workflowDescriptor: EngineWorkflowDescriptor,
         val conditionalStatus = if (b.value) ExecutionStatus.Done else ExecutionStatus.Bypassed
         val result = WorkflowExecutionDiff(conditionalKey.populate(workflowDescriptor.knownValues) + (conditionalKey -> conditionalStatus))
         result
-      case v: WdlValue => throw new RuntimeException("'if' condition must evaluate to a boolean")
+      case _: WdlValue => throw new RuntimeException("'if' condition must evaluate to a boolean")
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: TOL it'd be nice if we could say `... Boolean but instead got ${v.wdlType.toWdlString}`

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala [hunk 6]
```diff
@@ -510,7 +516,7 @@ case class WorkflowExecutionActor(workflowDescriptor: EngineWorkflowDescriptor,
       scatterKey.scope.collection.evaluate(lookup, data.expressionLanguageFunctions) map {
         case a: WdlArray =>
           WorkflowExecutionDiff(scatterKey.populate(a.value.size, workflowDescriptor.knownValues) + (scatterKey -> ExecutionStatus.Done))
-        case v: WdlValue => throw new RuntimeException("Scatter collection must evaluate to an array")
+        case _: WdlValue => throw new RuntimeException("Scatter collection must evaluate to an array")
```
Comment (cjllanwarne) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: TOL Same here

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala (hunk 1 above)
Comment (kshakir) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: A couple things:

- The other jobs have _already_ started by the point we do a `Try.sequence`. It's too late to stop them from launching. The EJEAs are already off to the races.
- Even when we do detect that something went wrong in "1 of 2000" jobs that just started, what is the duty of _this_ method?

Before, the method was throwing an exception killing the whole actor. The workflow wasn't actually stopped, and was left in an infinite "running" state. But yes even-more jobs didn't start. 😉 

Elsewhere in this WEA is other code (that isn't 100% [imho](https://github.com/broadinstitute/cromwell/issues/2029)) that decides whether to fail-fast or fail-slow when a single job fails. Again imho it's _that_ code that should decide what to do about the individual job failures, which is why I chose to send out a message to self.

So, happy to tech talk this, but I currently believe this method should be messaging, and not throwing exceptions.

File: engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala (hunk 4 above)
Comment (kshakir) [inline engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowExecutionActor.scala]: Not green == not my git-blame

**Comment by mcovarr** [comment]
discussed in person :+1:

**Comment by Horneth** [comment]
👍

**Pull request merged**